File: LeetCodes/path-with-maximum-probability/6-28-2023, 8_53_34 AM/path-with-maximum-probability.cpp
// Link: https://leetcode.com/problems/path-with-maximum-probability

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: path-with-maximum-probability
 *  RUNTIME: 222 ms
 *  MEMORY: 68.9 MB
 *  DATE: 6-28-2023, 8:53:34 AM
 *
 */

// Solution:

class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& prob, int start, int end) {
        vector<vector<pair<int,double>>> graph(n);
        for(int i=0; i<edges.size(); i++) {
            graph[edges[i][0]].push_back({edges[i][1], prob[i]});
            graph[edges[i][1]].push_back({edges[i][0], prob[i]});
        }

        queue<pair<int,double>> q;
        vector<double> vis(n, 0);
    
        q.push({start, 1});
        vis[start] = true;

        while(!q.empty()) {
            int node = q.front().first;
            double currProb = q.front().second;
            q.pop();   

            for(auto &x: graph[node]) {
                int next = x.first;
                double newProb = currProb * x.second;
                if(newProb > vis[next]) {
                    vis[next] = newProb;
                    q.push({next, newProb});
                }
            }
        }
        return vis[end];
    }
};

File: LeetCodes/path-with-maximum-probability/8-27-2024, 9_39_44 AM/path-with-maximum-probability.java
// Link: https://leetcode.com/problems/path-with-maximum-probability

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: path-with-maximum-probability
 *  RUNTIME: 10 ms
 *  MEMORY: 56.5 MB
 *  DATE: 8-27-2024, 9:39:44 AM
 *
 */

// Solution:

class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        double[] maxProb = new double[n];
        maxProb[start_node] = 1.0;

        for (int i = 0; i < n - 1; i++) {
            boolean updated = false;
            for (int j = 0; j < edges.length; j++) {
                int u = edges[j][0];
                int v = edges[j][1];
                double prob = succProb[j];

                if (maxProb[u] * prob > maxProb[v]) {
                    maxProb[v] = maxProb[u] * prob;
                    updated = true;
                }
                if (maxProb[v] * prob > maxProb[u]) {
                    maxProb[u] = maxProb[v] * prob;
                    updated = true;
                }
            }
            if (!updated) break;
        }

        return maxProb[end_node];
    }
}
